pkg/reconciler/taskrun/dynamicpool.go

```diff
@@ -0,0 +1,144 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"time"
+)
+
+type DynamicHostPool struct {
+	CloudProvider cloud.CloudProvider
+	SshSecret     string
+	Platform      string
+	MaxInstances  int
+	Concurrency   int
+	MaxAge        time.Duration
+	InstanceTag   string
+}
+
+func (a DynamicHostPool) buildHostPool(r *ReconcileTaskRun, ctx context.Context, log *logr.Logger, instanceTag string) (*HostPool, int, error) {
+	ret := map[string]*Host{}
+	instances, err := a.CloudProvider.ListInstances(r.client, log, ctx, instanceTag)
+	if err != nil {
+		return nil, 0, err
+	}
+	oldInstanceCount := 0
+	for _, instTmp := range instances {
+		inst := instTmp
+		if inst.StartTime.Add(a.MaxAge).Before(time.Now()) {
+			// These are shut down on deallocation
+			oldInstanceCount++
+
+			idle, err := a.isHostIdle(r, ctx, string(inst.InstanceId))
+			if err == nil {
+				if idle {
+					log.Info("deallocating old instance", "instance", inst.InstanceId)
+					err = a.CloudProvider.TerminateInstance(r.client, log, ctx, inst.InstanceId)
+					if err != nil {
+						log.Error(err, "unable to shut down instance", "instance", inst.InstanceId)
+					}
+				}
+			}
+		} else {
+			ret[string(inst.InstanceId)] = &Host{Name: string(inst.InstanceId), Address: inst.Address, User: a.CloudProvider.SshUser(), Concurrency: a.Concurrency, Platform: a.Platform, Secret: a.SshSecret, StartTime: &inst.StartTime}
+		}
+	}
+	return &HostPool{hosts: ret}, oldInstanceCount, nil
+}
+
+func (a DynamicHostPool) Deallocate(r *ReconcileTaskRun, ctx context.Context, log *logr.Logger, tr *v1.TaskRun, secretName string, selectedHost string) error {
+
+	hostPool, oldInstanceCount, err := a.buildHostPool(r, ctx, log, a.InstanceTag)
+	if err != nil {
+		return err
+	}
+	err = hostPool.Deallocate(r, ctx, log, tr, secretName, selectedHost)
+	if err != nil {
+		return err
+	}
+	if oldInstanceCount > 0 {
+		// Maybe this is an old instance
+		startTime := hostPool.hosts[selectedHost].StartTime
+		if startTime.Add(a.MaxAge).Before(time.Now()) {
+			// Old host, check if other tasks are using it
+			idle, err := a.isHostIdle(r, ctx, selectedHost)
+			if err != nil {
+				return err
+			}
+			if idle {
+				log.Info("deallocating old instance", "instance", selectedHost)
+				err = a.CloudProvider.TerminateInstance(r.client, log, ctx, cloud.InstanceIdentifier(selectedHost))
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (a DynamicHostPool) isHostIdle(r *ReconcileTaskRun, ctx context.Context, selectedHost string) (bool, error) {
+	trs := v1.TaskRunList{}
+	err := r.client.List(ctx, &trs, client.MatchingLabels{AssignedHost: selectedHost})
+	if err != nil {
+		return false, err
+	}
+	return len(trs.Items) == 0, nil
+}
+
+func (a DynamicHostPool) Allocate(r *ReconcileTaskRun, ctx context.Context, log *logr.Logger, tr *v1.TaskRun, secretName string, instanceTag string) (reconcile.Result, error) {
```

Comment:
When is this called? When a new TaskRun is created that requests a dynamic provision? Will the allocated VM just go to the back of the queue and the one from the front of the queue pop off?

Reply:
This is called when a task requests a VM. There is no queue as such, hosts are allocated based on current load. 

This implementation is kinda stateless, in that the state is maintained by either the cloud provider or the cluster.